### ISSUE TITLE
TF is unable to manage downstream EKS cluster resource w/ `desired_size=0, max_size=1, min_size=0`

### ISSUE THREAD

**Issue by Josh-Diamond** [rancher/terraform-provider-rancher2#989]
When provisioning a downstream EKS cluster via RancherUI, if `desired_size=0, max_size=1, min_size=0`, cluster will provision until it reaches a `Waiting` state, at which point it will remain in that state until nodes are added.  This is expected behavior.

With TF rancher2 provider, when attempting the scenario outlined above, once cluster reaches `Waiting` state, where it will remain until nodes are added, TF reports cluster resource as `still creating`.  This will not resolve.

When manually stopping TF from continuing its execution, modifying main.tf to reflect desired nodes, and running terraform apply again, this results in an error: `cluster already exists`.

It is impossible at this point for TF to manage the resource it has created.  This issue was created to address this circumstance.

At this point, cluster will only come up active by scaling up nodes through Rancher, without the use of TF.

**Additional Context**
Terraform is successful when initially deploying cluster w/ nodes, then scaling down to 0 nodes, and back up.  No issues, expected behavior.

Issue is only experienced when initially attempting to provision EKS w/ 0 nodes.

Screenshots:
<img width="951" alt="187995647-22d9314e-fbd8-4660-b66c-9ea278c74293" src="https://user-images.githubusercontent.com/46494969/189245123-911e2c7a-4f62-4ca9-b642-355543f63f5e.png">

---

It is my impression, that under this circumstance, allowing TF to recognize this `Waiting` state as the end of its resource creation, could address this behavior and resolve the issue.

**Comment by a-blender** [comment]
@Josh-Diamond Are we sure this is incorrect behavior?

As I understand it, when provisioning an EKS cluster (AWS hosted Kubernetes on EC2 nodes, basically) if the desiredSize = 0, then a node group will not be created until that value is scaled up. Tf is creating a downstream cluster and the cluster hasn't finished provisioning if there are no nodes so the "Still creating" message makes sense. I'm not sure I see the point of getting Terraform to think resource creation is complete without a kubeconfig. If a cluster is provisioned successfully to the point Terraform can manage it, then that is the relevant case. Why a customer would be doing this, is beyond me personally. Why is it in the QA test plan?

**Comment by Josh-Diamond** [comment]
I wouldn't say the behavior is incorrect, but I also would expect terraform to be able to manage a cluster resource that it began to  provision w/ 0 nodes, and have the option to scale.

When provisioning a downstream EKS w/ 0 nodes through Rancher, the cluster will stay in a `Waiting` state until nodes are added.  This may be done at anytime after initial provisioning by navigating to  `Edit Config` via Cluster Management. 

With terraform, this functionality doesn't currently exist, and so this ticket was created to (potentially) address this behavior.
Terraform will be able to manage a downstream cluster resource, if initially successfully provisioned.  You may scale down to 0 nodes, and back up if desired.  If attempting to initially provision w/ 0 nodes, terraform gets caught in a `still creating` loop, where it will never resolve and is unable to manage the resource it began to create.

If terraform is waiting for a valid kubeconfig to finish its resource creation, then perhaps the usecase this issue was created for is invalid/out-of-scope, and the expected behavior is already seen.

If this is expected behavior, would it be possible to add validation to prevent cluster provisioning if no nodes are defined in the config? (to prevent users from entering this loop, which never resolves?)

Also FWIW, this issue was not reported by a customer, but the result of exploratory testing on my part.

**Comment by a-blender** [comment]
Waiting on team discussion to give feedback on expected behavior i.e. is the "Still creating" loop expected behavior in Terraform for a cluster with 0 nodes? Then, we can decide how to address.

**Comment by a-blender** [comment]
**Design discussion summary**

Rancher currently is able to provision an EKS downstream cluster with 0 nodes, but hangs in a "Waiting state" and is _not_ active until a user scales up one or more node groups. This happens because there are no nodes for Rancher to deploy a control plane and an agent on.

Terraform, when provisioning with 0 nodes, hangs in a `Still creating...` state which _is expected behavior_ at this point because 1) the EKS new feature is so new, 2) the cluster resource has not returned a `kubeconfig` yet which Terraform is waiting for. Terraform is basically saying, "I can't manage this resource because it is not fully active" which is expected since that's what Rancher is also doing.

Our consensus is for Terraform to not try to create the cluster resources if the node group `desiredSize` is 0 (see [docs](https://registry.terraform.io/providers/rancher/rancher2/latest/docs/resources/cluster)). There needs to be code validation added to

a. Not create the resource here https://github.com/rancher/terraform-provider-rancher2/blob/afc19cc393199f93ec26814fbc8bbf7cc4cd6fb9/rancher2/resource_rancher2_cluster.go#L119
b. If the node group `desiredSize` is 0 https://github.com/rancher/terraform-provider-rancher2/blob/afc19cc393199f93ec26814fbc8bbf7cc4cd6fb9/rancher2/schema_cluster_eks_config_v2.go#L50

Assigned to @pgonin's team since this is KEv2 related.

**Comment by richardcase** [comment]
I can take a look at this

**Comment by mjura** [comment]
I have submitted fix proposal https://github.com/rancher/terraform-provider-rancher2/pull/1126

**Comment by a-blender** [comment]
@mjura Please be careful not to close issues unless they've been tested by QA! This needs a test template and will be available to test in RC v3.1.0-rc1.

**Comment by Josh-Diamond** [comment]
Ticket #989 - Test Results - ✅ 
---

With TFP `v3.1.0` on HA Helm env running Rancher `v2.7-027246f77f03b82660dc2e91df6bf2cd549163f0-head`:
1. Attempt to provision a downstream EKS cluster w/ `desired_size` of `node_groups` set to `0`
2. Verified - validation prevents this invalid use case from occurring; as expected

Screenshot:

<img width="736" alt="Screenshot 2023-07-28 at 4 26 57 PM" src="https://github.com/rancher/terraform-provider-rancher2/assets/46494969/e516e29e-3676-41b1-b4b8-3371c8910590">

---

Note: This ticket is validated and as expected so closing this out, but will open a new enhancement against this, to improve the error message.  Suggested error message below:

```
[ERROR] cannot create %s EKS cluster with node group desired_size = 0.  desired_size must be >=1. After initial provisioning, desired_size may be scaled down to 0 at any time.
```
